src/Microsoft.AspNet.BasicMiddleware/OverrideHeaderMiddleware.cs

```diff
@@ -0,0 +1,76 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Net;
+using System.Threading.Tasks;
+using Microsoft.AspNet.Builder;
+using Microsoft.AspNet.Http;
+
+namespace Microsoft.AspNet.BasicMiddleware
+{
+    public class OverrideHeaderMiddleware
+    {
+        private const string XForwardedFor = "X-Forwarded-For";
+        private const string XForwardedHost = "X-Forwarded-Host";
+        private const string XForwardedProto = "X-Forwarded-Proto";
+        private readonly OverrideHeaderMiddlewareOptions _options;
+        private readonly RequestDelegate _next;
+
+
+        public OverrideHeaderMiddleware(RequestDelegate next, OverrideHeaderMiddlewareOptions options)
+        {
+            if (options == null || next == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+            _options = options;
+            _next = next;
+        }
+
+        /// <summary>
+        ///  Read the values for the different header keys we are looking for
+        /// then overrode those values in the request with the values in the header
+        ///     - X-Forward-For
+        ///     - X-Forwarded-Host
+        ///     - X-Forwarded-Proto
+        /// </summary>
+        /// <param name="context"></param>
+        /// <returns></returns>
+        public Task Invoke(HttpContext context)
+        {
+            if (_options.XForwardedForEnabled)
+            {
+                var xForward = context.Request.Headers.GetCommaSeparatedValues(XForwardedFor);
+                if (xForward != null && xForward.Count > 0)
+                {
+                    IPAddress originalIPAddress;
+                    if (IPAddress.TryParse(xForward[0], out originalIPAddress))
```

Comment:
Is the expected behavior that you forward the first ip address? What if there is more than one?